lib/SILOptimizer/FunctionSignatureTransforms/ExistentialSpecializer.cpp

```diff
@@ -0,0 +1,359 @@
+//===--- ExistentialSpecializer.cpp - Specialization of functions -----===//
+//===---                 with existential arguments               -----===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Specialize functions with existential parameters to generic ones.
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "sil-existential-specializer"
+#include "ExistentialTransform.h"
+#include "swift/SIL/SILFunction.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/Existential.h"
+#include "swift/SILOptimizer/Utils/Local.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/Statistic.h"
+
+using namespace swift;
+
+STATISTIC(NumFunctionsWithExistentialArgsSpecialized,
+          "Number of functions with existential args specialized");
+
+namespace {
+
+/// ExistentialSpecializer class.
+class ExistentialSpecializer : public SILFunctionTransform {
+
+  /// Determine if the current function is a target for existential
+  /// specialization of args.
+  bool canSpecializeExistentialArgsInFunction(
+      ApplySite &Apply,
+      llvm::SmallDenseMap<int, ExistentialTransformArgumentDescriptor>
```

Comment:
Could this just be a SmallVector? Your indices are in order and contiguous

Reply:
Ping?

Reply:
An ExistentialTransformArgumentDescriptor is only built for an existential argument, not other arguments. So,SmallVector is not the right option.

Reply:
This is fine?